iot_verifier/src/rewarder.rs

```diff
@@ -141,3 +274,108 @@ async fn save_rewarded_timestamp(
 ) -> db_store::Result<()> {
     meta::store(db, timestamp_key, value.timestamp()).await
 }
+
+pub async fn fetch_halt_rewards_override(db: impl sqlx::PgExecutor<'_>) -> db_store::Result<bool> {
+    Ok(sqlx::query_scalar::<_, bool>(
+        r#"
+    SELECT EXISTS(SELECT value from meta where key = $1 and value = 'true')
+    "#,
+    )
+    .bind(HALT_REWARDS_DB_KEY)
+    .fetch_one(db)
+    .await?)
+}
+
+async fn clear_halt_rewards_override(
+    txn: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+) -> db_store::Result<()> {
+    meta::store(txn, HALT_REWARDS_DB_KEY, "false").await
+}
+
+async fn save_history(
+    txn: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+    total_beacon_count: u64,
+    total_witness_count: u64,
+    total_packet_count: u64,
+    epoch_end: DateTime<Utc>,
+) -> anyhow::Result<()> {
+    history_db::insert(
+        txn,
+        RewardType::Beacon,
+        total_beacon_count as i64,
+        epoch_end,
+    )
+    .await?;
+    history_db::insert(
+        txn,
+        RewardType::Witness,
+        total_witness_count as i64,
+        epoch_end,
+    )
+    .await?;
+    history_db::insert(
+        txn,
+        RewardType::Packet,
+        total_packet_count as i64,
+        epoch_end,
+    )
+    .await?;
+    Ok(())
+}
+
+mod history_db {
+    use super::*;
+
+    pub async fn insert(
+        tx: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+        reward_type: RewardType,
+        count: i64,
+        epoch_ts: DateTime<Utc>,
+    ) -> anyhow::Result<()> {
+        sqlx::query(
+            r#"
+        INSERT INTO rewards_history(reward_type, count, epoch_ts) VALUES($1, $2, $3)
+        "#,
+        )
+        .bind(reward_type)
+        .bind(count)
+        .bind(epoch_ts)
+        .execute(tx)
+        .await?;
```

Comment:
```suggestion
        beacon_count: i64,
        witness_count: i64,
        packet_count: i64,
        epoch_ts: DateTime<Utc>,
    ) -> anyhow::Result<()> {
        sqlx::query(
            r#"
        INSERT INTO rewards_history (reward_type, count, epoch_ts) VALUES ('beacon', $1, $4), ('witness', $2, $4), ('packet', $3, $4)
        "#,
        )
        .bind(beacon_count)
        .bind(witness_count)
        .bind(packet_count)
        .bind(epoch_ts)
        .execute(tx)
        .await?;
```
since this is a fixed operation you could do this with a single DB call